File: packages/types/src/document.ts
export type IngestionStatus = "queued" | "processing" | "processed" | "failed";

export interface DocumentMetadata {
  title: string;
  author?: string;
  pageCount?: number;
  fileSize: number;
  mimeType: string;
  checksum: string;
}

export interface Document {
  id: string;
  name: string;
  storageUrl: string;
  status: "queued" | "UPLOADING" | "PARSING" | "CLEANING" | "CHUNKING" | "EMBEDDING" | "INDEXING" | "ANALYZING" | "COMPLETED" | "FAILED" | "processing" | "processed" | "failed";
  metadata: DocumentMetadata;
  createdAt: string;
  updatedAt: string;
  progress_percentage?: number;
  userId?: string;
  error?: string | null;
}

export interface UploadParams {
  name: string;
  fileSize: number;
  mimeType: string;
}

export interface IngestDocumentResponse {
  documentId: string;
  status: IngestionStatus;
  uploadUrl?: string; // Pre-signed S3/R2 upload URL
}

File: apps/api/schemas/document.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict
from enum import Enum

class IngestionStatus(str, Enum):
    QUEUED = "queued"
    UPLOADING = "UPLOADING"
    PARSING = "PARSING"
    CLEANING = "CLEANING"
    CHUNKING = "CHUNKING"
    EMBEDDING = "EMBEDDING"
    INDEXING = "INDEXING"
    ANALYZING = "ANALYZING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    # Legacy fallbacks
    PROCESSING = "processing"
    PROCESSED = "processed"
    failed = "failed"

class DocumentMetadataSchema(BaseModel):
    title: str
    author: str | None = None
    pageCount: int | None = None
    fileSize: int
    mimeType: str
    checksum: str

class DocumentBase(BaseModel):
    name: str

class DocumentResponse(DocumentBase):
    id: str
    storageUrl: str
    status: IngestionStatus
    metadata: DocumentMetadataSchema
    createdAt: datetime
    updatedAt: datetime
    userId: str
    progress_percentage: int = 0
    error: str | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

class UploadParams(BaseModel):
    name: str
    fileSize: int
    mimeType: str
